src/ol/format/esrijsonformat.js

```diff
@@ -0,0 +1,687 @@
+ */
+ol.format.EsriJSON.writePolygonGeometry_ = function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.Polygon,
+      'geometry should be an ol.geom.Polygon');
+  // Esri geometries use the left-hand rule
+  var hasZM = ol.format.EsriJSON.getHasZM_(geometry);
+  return /** @type {EsriJSONPolygon} */ ({
+    'hasZ': hasZM.hasZ,
+    'hasM': hasZM.hasM,
+    'rings': geometry.getCoordinates(false)
+  });
+};
+
+
+/**
+ * @param {ol.geom.Geometry} geometry Geometry.
+ * @param {olx.format.WriteOptions=} opt_options Write options.
+ * @private
+ * @return {EsriJSONPolyline} EsriJSON geometry.
+ */
+ol.format.EsriJSON.writeMultiLineStringGeometry_ =
+    function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.MultiLineString,
+      'geometry should be an ol.geom.MultiLineString');
+  var hasZM = ol.format.EsriJSON.getHasZM_(geometry);
+  return /** @type {EsriJSONPolyline} */ ({
+    'hasZ': hasZM.hasZ,
+    'hasM': hasZM.hasM,
+    'paths': geometry.getCoordinates()
+  });
+};
+
+
+/**
+ * @param {ol.geom.Geometry} geometry Geometry.
+ * @param {olx.format.WriteOptions=} opt_options Write options.
+ * @private
+ * @return {EsriJSONMultipoint} EsriJSON geometry.
+ */
+ol.format.EsriJSON.writeMultiPointGeometry_ = function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.MultiPoint,
+      'geometry should be an ol.geom.MultiPoint');
+  var hasZM = ol.format.EsriJSON.getHasZM_(geometry);
+  return /** @type {EsriJSONMultipoint} */ ({
+    'hasZ': hasZM.hasZ,
+    'hasM': hasZM.hasM,
+    'points': geometry.getCoordinates()
+  });
+};
+
+
+/**
+ * @param {ol.geom.Geometry} geometry Geometry.
+ * @param {olx.format.WriteOptions=} opt_options Write options.
+ * @private
+ * @return {EsriJSONPolygon} EsriJSON geometry.
+ */
+ol.format.EsriJSON.writeMultiPolygonGeometry_ = function(geometry,
+    opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.MultiPolygon,
+      'geometry should be an ol.geom.MultiPolygon');
+  var hasZM = ol.format.EsriJSON.getHasZM_(geometry);
+  var coordinates = geometry.getCoordinates(false);
+  var output = [];
+  for (var i = 0; i < coordinates.length; i++) {
+    for (var x = coordinates[i].length - 1; x >= 0; x--) {
+      output.push(coordinates[i][x]);
+    }
+  }
+  return /** @type {EsriJSONPolygon} */ ({
+    'hasZ': hasZM.hasZ,
+    'hasM': hasZM.hasM,
+    'rings': output
+  });
+};
+
+
+/**
+ * @const
+ * @private
+ * @type {Object.<string, function(EsriJSONObject): ol.geom.Geometry>}
+ */
+ol.format.EsriJSON.GEOMETRY_READERS_ = {
+  'Point': ol.format.EsriJSON.readPointGeometry_,
+  'LineString': ol.format.EsriJSON.readLineStringGeometry_,
+  'Polygon': ol.format.EsriJSON.readPolygonGeometry_,
+  'MultiPoint': ol.format.EsriJSON.readMultiPointGeometry_,
+  'MultiLineString': ol.format.EsriJSON.readMultiLineStringGeometry_,
+  'MultiPolygon': ol.format.EsriJSON.readMultiPolygonGeometry_
```

Comment:
These don't actually have to be quoted keys.  In GeoJSON, the `type` comes directly from the string we are reading, so we need to compare against quoted strings.  Here, I think you could use the `ol.geom.GeometryType` enum.

In calling code you would do something like this:

```js
  var type = ol.geom.Geometry.POINT;
  var reader = ol.format.EsriJSON.GEOMETRY_READERS_[type];
```

And then this object would be created with those enum values.  E.g.
```js
ol.format.EsriJSON.GEOMETRY_READERS_ = {};
ol.format.EsriJSON.GEOMETRY_READERS_[ol.geom.GeometryType.POINT] = 
    ol.format.EsriJSON.readPointGeometry_;
// ...
```